textutils-2.1/man/uniq.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.28.
.TH UNIQ "1" "July 2002" "uniq (textutils) 2.1" "User Commands"
.SH NAME
uniq \- remove duplicate lines from a sorted file
.SH SYNOPSIS
.B uniq
[\fIOPTION\fR]... [\fIINPUT \fR[\fIOUTPUT\fR]]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Discard all but one of successive identical lines from INPUT (or
standard input), writing to OUTPUT (or standard output).
.PP
Mandatory arguments to long options are mandatory for short options too.
.TP
\fB\-c\fR, \fB\-\-count\fR
prefix lines by the number of occurrences
.TP
\fB\-d\fR, \fB\-\-repeated\fR
only print duplicate lines
.TP
\fB\-D\fR, \fB\-\-all\-repeated\fR[=\fIdelimit\-method\fR] print all duplicate lines
delimit-method={none(default),prepend,separate}
Delimiting is done with blank lines.
.TP
\fB\-f\fR, \fB\-\-skip\-fields\fR=\fIN\fR
avoid comparing the first N fields
.TP
\fB\-i\fR, \fB\-\-ignore\-case\fR
ignore differences in case when comparing
.TP
\fB\-s\fR, \fB\-\-skip\-chars\fR=\fIN\fR
avoid comparing the first N characters
.TP
\fB\-u\fR, \fB\-\-unique\fR
only print unique lines
.TP
\fB\-w\fR, \fB\-\-check\-chars\fR=\fIN\fR
compare no more than N characters in lines
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-version\fR
output version information and exit
.PP
A field is a run of whitespace, then non-whitespace characters.
Fields are skipped before chars.
.SH AUTHOR
Written by Richard Stallman and David MacKenzie.
.SH "REPORTING BUGS"
Report bugs to <bug-textutils@gnu.org>.
.SH COPYRIGHT
Copyright \(co 2002 Free Software Foundation, Inc.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
The full documentation for
.B uniq
is maintained as a Texinfo manual.  If the
.B info
and
.B uniq
programs are properly installed at your site, the command
.IP
.B info uniq
.PP
should give you access to the complete manual.


zsh/ChangeLog
beta 0.75
Wed Jun 14 14:47:46 2000    <amold@SKYNYRD>
	* ntport\dirent.c: Fix serious bug in wildcard expansion on Win9x 
beta 0.74
Mon Jan 31 14:53:06 2000    <amold@SKYNYRD>
	* ntport\support.c: fixed bug in converting '\'s in $HOME 
	to '/'s.
beta 0.73
Mon Jan 17 11:55:15 2000    <amold@SKYNYRD>
	* ntport\support.c: fixed crash in long arglist introduced by
 	0.70
beta 0.72
Tue Jan 11 08:28:29 2000    <amold@SKYNYRD>
	* ntport\support.c: Fixed memory corruption on Win9x
beta 0.71
Mon Jan 10 11:49:39 2000    <amold@SKYNYRD>
	* ntport\stdio.c: Fix bug in inheritance of file handles.  
beta 0.69,0.70
Fri Dec 17 11:53:19 1999    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug in argv[0] of spawned argument. Might
	have manifested as applications in ZSHROOT failing to spawn 
	themselves. 
	Also added winntnoquoteprotect option. Use with tools like MKS
	if you see unexpected quoting of "
beta 0.68
Mon Sep 27 11:20:24 1999    <amold@SKYNYRD>
	* ntport\support.c: support paths > 1024 characters. 
beta 0.67
Tue Sep 07 10:05:02 1999    <amold@SKYNYRD>
	* ntport\io.c: Fix bug that affected FPATH/autoload 
beta 0.65
Fri Aug 27 11:41:58 1999    <amold@SKYNYRD>
	* ntport\dirent.c: Fix completion with drive letters in string. 
beta 0.64
Fri Aug 27 11:04:11 1999    <amold@SKYNYRD>
	* ntport\io.c: Unfix Win95 spurious linefeed fix for Windows 98. 
beta 0.63
Wed May 19 16:21:37 1999    <amold@SKYNYRD>
	* exec.c, misc: VC 6.0 compliation errors. Disable optmizations
	in functions that call fork() to avoid losing global variables
	that might be in registers.
beta 0.62
Tue Apr 13 17:06:12 1999    <amold@SKYNYRD>
	* ntport\support.c, zle_misc.c: Keep server name in prompt
 	if %C is specified and pwd is UNC.
beta 0.61
Tue Mar 23 12:17:50 1999  unknown  <bogus@SKYNYRD>
	* ntport\support.c: bug fix for previous fix. 
beta 0.60
Sat Feb 13 17:49:14 1999    <amold@SKYNYRD>
	* ntport\support.c,ntbin.c: Fix lame arg-concatenation in
 	exec and start paths. commands with large argument strings should
	now execute *much* faster.
beta 0.57
Tue Feb 02 18:09:39 1999    <amold@SKYNYRD>
	* ntport\ntbin.c: fixed bug that improperly passed arguments
	to applications launched by windows "shell" associations. 
	For example, .bat files.
beta 0.55
Mon Oct 26 11:09:44 1998    <amold@SKYNYRD>
	* ntport\ntbin.c: start builtin will not use '/' for option
 	flags. only '-' will work.
beta 0.54
Fri Oct 16 12:52:08 1998    <amold@SKYNYRD>
	* ntport\support.c: added WINNTWAITFORGUIAPPS option. 
beta 0.53
Mon Oct 12 12:31:59 1998    <amold@SKYNYRD>
	* init.c: look for "zshrc" as well as ".zshrc" 
beta 0.52
Mon Sep 28 17:31:08 1998    <amold@SKYNYRD>
	* utils.c: attempt to ignore case while correcting spelling
 	if WINNTIGNORECASE is set.
beta 0.51
Thu Sep 17 13:48:27 1998    <amold@SKYNYRD>
	* hashtable.c,support.c: filenames without extensions and
 	*only* files with extensions specified in the PATHEXT variable
 	will be hashed. The default extensions are ".com;.exe;.cmd;.bat"
beta 0.50
Wed Sep 16 15:58:56 1998    <amold@SKYNYRD>
	* hashtable.c: added hashing for external exes without
	extension. bogus spelling corrections should be reduced.
beta 0.49
Tue Sep 15 11:09:31 1998    <amold@SKYNYRD>
	* support.c: The shell will automatically set TERM to vt100
 	if it's not set in the environment. There should be no need
	to set the variable explicitly any more.
beta 0.48
Mon Aug 03 17:47:08 1998    <amold@SKYNYRD>
	* support.c: Fixed bug in converting paths to '/'-delimited 
beta 0.47
Mon Aug 03 16:59:38 1998    <amold@SKYNYRD>
	* utils.c: Fixed bug in applying previous patch. If %~ wasn't
	working, this should fix it.
beta 0.46
Sat Jul 18 14:36:10 1998    <amold@SKYNYRD>
	* utils.c,globals.h,zsh.h: patch from dominik for 
	winntconvertbackslash.
beta 0.45
Mon Jun 29 11:30:55 1998    <amold@SKYNYRD>
	* utils.c: patch from dominik vogt for honoring winntignorecase
 	in %~ prompt.
beta 0.44
Mon Jun 22 11:19:12 1998    <amold@SKYNYRD>
	* support.c,io.c: moved some startup code around. fixed spurious
 	linefeeds on win95
beta 0.43
Tue Jun 09 17:25:29 1998    <amold@SKYNYRD>
	* builtin.c,ntport\io.c: Fix for the read builtin not
 	echoing characters
beta 0.429
Thu Jun 04 17:43:54 1998    <amold@SKYNYRD>
	* utils.c: Fixed %~ prompt spec.
beta 0.428
Thu Jun 04 16:50:21 1998    <amold@SKYNYRD>
	* builtin.c: Fixed problems with the setting of PWD.
beta 0.427
Thu May 28 14:31:28 1998    <amold@SKYNYRD>
	* ntport\support.c: Fix to let apps with spaces in their path
	run. c:\program files\myapp.exe , for example.
beta 0.426
Wed May 27 10:46:10 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed stupid command buffer overflow
 	problem. If you crashed while grepping through a bunch of
	files, this should fix it.
beta 0.425
Fri May 22 08:55:10 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug in parsing args to child processes.
beta 0.424
Sun May 17 13:18:21 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug that caused zsh to spawn a new
	console if $path contained a component '/'. 
beta 0.423
Fri May 15 16:08:17 1998    <amold@SKYNYRD>
	* ntport\ntbin.c: winntlamepathfix now works with the 'start'
	builtin.
beta 0.421
Mon Apr 06 16:18:20 1998    <amold@SKYNYRD>
	* ntport\support.c,fork.c: Fixed bug in heap initialization for
	forked children that crashed the child process.
beta 0.42
Sun Apr 05 13:36:49 1998    <amold@SKYNYRD>
	* ntport\support.c: Put in code to fix the case of a directory
 	you cd into. This is extremely expensive and really silly, so
	it may become optional in the future.
beta 0.41
Sat Apr 04 15:27:10 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug that restricted command line
 	length
beta 0.401
Thu Mar 26 08:21:21 1998    <amold@SKYNYRD>
	* ntport\support.c: Fix to stop home from being re-set in child
 	if unset in parent. Also HOME will not be set by zsh if ZDOTDIR

textutils-2.1/src/uniq.c
/* uniq -- remove duplicate lines from a sorted file
   Copyright (C) 86, 91, 1995-2002, Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

/* Written by Richard Stallman and David MacKenzie. */

#include <config.h>

#include <stdio.h>
#include <getopt.h>
#include <sys/types.h>

#include "system.h"
#include "closeout.h"
#include "argmatch.h"
#include "linebuffer.h"
#include "error.h"
#include "posixver.h"
#include "xstrtol.h"
#include "memcasecmp.h"

/* The official name of this program (e.g., no `g' prefix).  */
#define PROGRAM_NAME "uniq"

#define AUTHORS N_ ("Richard Stallman and David MacKenzie")

#define SWAP_LINES(A, B)			\
  do						\
    {						\
      struct linebuffer *_tmp;			\
      _tmp = (A);				\
      (A) = (B);				\
      (B) = _tmp;				\
    }						\
  while (0)

/* The name this program was run with. */
char *program_name;

/* Number of fields to skip on each line when doing comparisons. */
static size_t skip_fields;

/* Number of chars to skip after skipping any fields. */
static size_t skip_chars;

/* Number of chars to compare. */
static size_t check_chars;

enum countmode
{
  count_occurrences,		/* -c Print count before output lines. */
  count_none			/* Default.  Do not print counts. */
};

/* Whether and how to precede the output lines with a count of the number of
   times they occurred in the input. */
static enum countmode countmode;

enum output_mode
{
  output_repeated,		/* -d Only lines that are repeated. */
  output_all_repeated,		/* -D All lines that are repeated. */
  output_unique,		/* -u Only lines that are not repeated. */
  output_all			/* Default.  Print first copy of each line. */
};

/* Which lines to output. */
static enum output_mode mode;

/* If nonzero, ignore case when comparing.  */
static int ignore_case;

enum delimit_method
{
  /* No delimiters output.  --all-repeated[=none] */
  DM_NONE,

  /* Delimiter precedes all groups.  --all-repeated=prepend */
  DM_PREPEND,

  /* Delimit all groups.  --all-repeated=separate */
  DM_SEPARATE
};

static char const *const delimit_method_string[] =
{
  "none", "prepend", "separate", 0
};

static enum delimit_method const delimit_method_map[] =
{
  DM_NONE, DM_PREPEND, DM_SEPARATE
};

/* Select whether/how to delimit groups of duplicate lines.  */
static enum delimit_method delimit_groups;

static struct option const longopts[] =
{
  {"count", no_argument, NULL, 'c'},
  {"repeated", no_argument, NULL, 'd'},
  {"all-repeated", optional_argument, NULL, 'D'},
  {"ignore-case", no_argument, NULL, 'i'},
  {"unique", no_argument, NULL, 'u'},
  {"skip-fields", required_argument, NULL, 'f'},
  {"skip-chars", required_argument, NULL, 's'},
  {"check-chars", required_argument, NULL, 'w'},
  {GETOPT_HELP_OPTION_DECL},
  {GETOPT_VERSION_OPTION_DECL},
  {NULL, 0, NULL, 0}
};

void
usage (int status)
{
  if (status != 0)
    fprintf (stderr, _("Try `%s --help' for more information.\n"),
	     program_name);
  else
    {
      printf (_("\
Usage: %s [OPTION]... [INPUT [OUTPUT]]\n\
"),
	      program_name);
      fputs (_("\
Discard all but one of successive identical lines from INPUT (or\n\
standard input), writing to OUTPUT (or standard output).\n\
\n\
"), stdout);
     fputs (_("\
Mandatory arguments to long options are mandatory for short options too.\n\
"), stdout);
     fputs (_("\
  -c, --count           prefix lines by the number of occurrences\n\
  -d, --repeated        only print duplicate lines\n\
"), stdout);
     fputs (_("\
  -D, --all-repeated[=delimit-method] print all duplicate lines\n\
                        delimit-method={none(default),prepend,separate}\n\
                        Delimiting is done with blank lines.\n\
  -f, --skip-fields=N   avoid comparing the first N fields\n\
  -i, --ignore-case     ignore differences in case when comparing\n\
  -s, --skip-chars=N    avoid comparing the first N characters\n\
  -u, --unique          only print unique lines\n\
"), stdout);
     fputs (_("\
  -w, --check-chars=N   compare no more than N characters in lines\n\
"), stdout);
     fputs (HELP_OPTION_DESCRIPTION, stdout);
     fputs (VERSION_OPTION_DESCRIPTION, stdout);
     fputs (_("\
\n\
A field is a run of whitespace, then non-whitespace characters.\n\
Fields are skipped before chars.\n\
"), stdout);
      printf (_("\nReport bugs to <%s>.\n"), PACKAGE_BUGREPORT);
    }
  exit (status == 0 ? EXIT_SUCCESS : EXIT_FAILURE);
}

/* Convert OPT to size_t, reporting an error using MSGID if it does
   not fit.  */

static size_t
size_opt (char const *opt, char const *msgid)
{
  unsigned long int size;
  if (xstrtoul (opt, NULL, 10, &size, "") != LONGINT_OK
      || SIZE_MAX < size)
    error (EXIT_FAILURE, 0, "%s: %s", opt, _(msgid));
  return size;
}

/* Given a linebuffer LINE,
   return a pointer to the beginning of the line's field to be compared. */

static char *
find_field (const struct linebuffer *line)
{
  register size_t count;
  register char *lp = line->buffer;
  register size_t size = line->length - 1;
  register size_t i = 0;

  for (count = 0; count < skip_fields && i < size; count++)
    {
      while (i < size && ISBLANK (lp[i]))
	i++;
      while (i < size && !ISBLANK (lp[i]))
	i++;
    }

  for (count = 0; count < skip_chars && i < size; count++)
    i++;

  return lp + i;
}

/* Return zero if two strings OLD and NEW match, nonzero if not.
   OLD and NEW point not to the beginnings of the lines
   but rather to the beginnings of the fields to compare.
   OLDLEN and NEWLEN are their lengths. */

static int
different (const char *old, const char *new, size_t oldlen, size_t newlen)
{
  if (check_chars < oldlen)
    oldlen = check_chars;
  if (check_chars < newlen)
    newlen = check_chars;

  if (oldlen != newlen)
    return 1;

  /* Use an if-statement here rather than a function variable to
     avoid portability hassles of getting a non-conflicting declaration
     of memcmp.  */
  if (ignore_case)
    return memcasecmp (old, new, oldlen);
  else
    return memcmp (old, new, oldlen);
}

/* Output the line in linebuffer LINE to stream STREAM
   provided that the switches say it should be output.
   If requested, print the number of times it occurred, as well;
   LINECOUNT + 1 is the number of times that the line occurred. */

static void
writeline (const struct linebuffer *line, FILE *stream, int linecount)
{
  if ((mode == output_unique && linecount != 0)
      || (mode == output_repeated && linecount == 0)
      || (mode == output_all_repeated && linecount == 0))
    return;

  if (countmode == count_occurrences)
    fprintf (stream, "%7d\t", linecount + 1);

  fwrite (line->buffer, sizeof (char), line->length, stream);
}

/* Process input file INFILE with output to OUTFILE.
   If either is "-", use the standard I/O stream for it instead. */

static void
check_file (const char *infile, const char *outfile)
{
  FILE *istream;
  FILE *ostream;
  struct linebuffer lb1, lb2;
  struct linebuffer *thisline, *prevline;

  if (STREQ (infile, "-"))
    istream = stdin;
  else
    istream = fopen (infile, "rb");
  if (istream == NULL)
    error (EXIT_FAILURE, errno, "%s", infile);

  if (STREQ (outfile, "-"))
    ostream = stdout;
  else
    ostream = fopen (outfile, "wb");
  if (ostream == NULL)
    error (EXIT_FAILURE, errno, "%s", outfile);

  thisline = &lb1;
  prevline = &lb2;

  initbuffer (thisline);
  initbuffer (prevline);

  /* The duplication in the following `if' and `else' blocks is an
     optimization to distinguish the common case (in which none of
     the following options has been specified: --count, -repeated,
     --all-repeated, --unique) from the others.  In the common case,
     this optimization lets uniq output each different line right away,
     without waiting to see if the next one is different.  */

  if (mode == output_all && countmode == count_none)
    {
      char *prevfield IF_LINT (= NULL);
      size_t prevlen IF_LINT (= 0);

      while (!feof (istream))
	{
	  char *thisfield;
	  size_t thislen;
	  if (readline (thisline, istream) == 0)
	    break;
	  thisfield = find_field (thisline);
	  thislen = thisline->length - (thisfield - thisline->buffer);
	  if (prevline->length == 0
	      || different (thisfield, prevfield, thislen, prevlen))
	    {
	      fwrite (thisline->buffer, sizeof (char),
		      thisline->length, ostream);

	      SWAP_LINES (prevline, thisline);
	      prevfield = thisfield;
	      prevlen = thislen;
	    }
	}
    }
  else
    {
      char *prevfield;
      size_t prevlen;
      int match_count = 0;
      int first_delimiter = 1;

      if (readline (prevline, istream) == 0)
	goto closefiles;
      prevfield = find_field (prevline);
      prevlen = prevline->length - (prevfield - prevline->buffer);

      while (!feof (istream))
	{
	  int match;
	  char *thisfield;
	  size_t thislen;
	  if (readline (thisline, istream) == 0)
	    break;
	  thisfield = find_field (thisline);
	  thislen = thisline->length - (thisfield - thisline->buffer);
	  match = !different (thisfield, prevfield, thislen, prevlen);

	  if (match)
	    ++match_count;

          if (mode == output_all_repeated && delimit_groups != DM_NONE)
	    {
	      if (!match)
		{
		  if (match_count) /* a previous match */
		    first_delimiter = 0; /* Only used when DM_SEPARATE */
		}
	      else if (match_count == 1)
		{
		  if ((delimit_groups == DM_PREPEND)
		      || (delimit_groups == DM_SEPARATE
			  && !first_delimiter))
		    putc ('\n', ostream);
		}
	    }

	  if (!match || mode == output_all_repeated)
	    {
	      writeline (prevline, ostream, match_count);
	      SWAP_LINES (prevline, thisline);
	      prevfield = thisfield;
	      prevlen = thislen;
	      if (!match)
		match_count = 0;
	    }
	}

      writeline (prevline, ostream, match_count);
    }

 closefiles:
  if (ferror (istream) || fclose (istream) == EOF)
    error (EXIT_FAILURE, errno, _("error reading %s"), infile);

  /* Close ostream only if it's not stdout -- the latter is closed
     via the atexit-invoked close_stdout.  */
  if (ostream != stdout && (ferror (ostream) || fclose (ostream) == EOF))
    error (EXIT_FAILURE, errno, _("error writing %s"), outfile);

  free (lb1.buffer);
  free (lb2.buffer);
}

int
main (int argc, char **argv)
{
  int optc = 0;
  bool posixly_correct = (getenv ("POSIXLY_CORRECT") != NULL);
  bool obsolete_skip_fields = false;
  int nfiles = 0;
  char const *file[2];

  file[0] = file[1] = "-";
  program_name = argv[0];
  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  atexit (close_stdout);

  skip_chars = 0;
  skip_fields = 0;
  check_chars = SIZE_MAX;
  mode = output_all;
  countmode = count_none;
  delimit_groups = DM_NONE;

  for (;;)
    {
      /* Parse an operand with leading "+" as a file after "--" was
         seen; or if pedantic and a file was seen; or if not
         obsolete.  */

      if (optc == -1
	  || (posixly_correct && nfiles != 0)
	  || ((optc = getopt_long (argc, argv,
				   "-0123456789Dcdf:is:uw:", longopts, NULL))
	      == -1))
	{
	  if (optind == argc)
	    break;
	  if (nfiles == 2)
	    {
	      error (0, 0, _("extra operand `%s'"), argv[optind]);
	      usage (1);
	    }
	  file[nfiles++] = argv[optind++];
	}
      else switch (optc)
	{
	case 1:
	  {
	    unsigned long int size;
	    if (optarg[0] == '+'
		&& posix2_version () < 200112
		&& xstrtoul (optarg, NULL, 10, &size, "") == LONGINT_OK
		&& size <= SIZE_MAX)
	      skip_chars = size;
	    else if (nfiles == 2)
	      {
		error (0, 0, _("extra operand `%s'"), optarg);
		usage (1);
	      }
	    else
	      file[nfiles++] = optarg;
	  }
	  break;

	case '0':
	case '1':
	case '2':
	case '3':
	case '4':
	case '5':
	case '6':
	case '7':
	case '8':
	case '9':
	  {
	    size_t s = skip_fields;
	    skip_fields = s * 10 + optc - '0';
	    if (SIZE_MAX / 10 < s || skip_fields < s)
	      error (EXIT_FAILURE, 0, "%s",
		     _("invalid number of fields to skip"));
	    obsolete_skip_fields = true;
	  }
	  break;

	case 'c':
	  countmode = count_occurrences;
	  break;

	case 'd':
	  mode = output_repeated;
	  break;

	case 'D':
	  mode = output_all_repeated;
	  if (optarg == NULL)
	    delimit_groups = DM_NONE;
	  else
	    delimit_groups = XARGMATCH ("--all-repeated", optarg,
					delimit_method_string,
					delimit_method_map);
	  break;

	case 'f':		/* Like '-#'. */
	  skip_fields = size_opt (optarg,
				  N_("invalid number of fields to skip"));
	  break;

	case 'i':
	  ignore_case = 1;
	  break;

	case 's':		/* Like '+#'. */
	  skip_chars = size_opt (optarg,
				 N_("invalid number of bytes to skip"));
	  break;

	case 'u':
	  mode = output_unique;
	  break;

	case 'w':
	  check_chars = size_opt (optarg,
				  N_("invalid number of bytes to compare"));
	  break;

	case_GETOPT_HELP_CHAR;

	case_GETOPT_VERSION_CHAR (PROGRAM_NAME, AUTHORS);

	default:
	  usage (1);
	}
    }

  if (obsolete_skip_fields && 200112 <= posix2_version ())
    {
      error (0, 0, _("`-%lu' option is obsolete; use `-f %lu'"),
	     (unsigned long) skip_fields, (unsigned long) skip_fields);
      usage (EXIT_FAILURE);
    }

  if (countmode == count_occurrences && mode == output_all_repeated)
    {
      error (0, 0,
	   _("printing all duplicated lines and repeat counts is meaningless"));
      usage (1);
    }

  check_file (file[0], file[1]);

  exit (EXIT_SUCCESS);
}
